packages/core/src/common/abstractComponent.ts

```diff
@@ -21,26 +21,39 @@ import { isNodeEnv } from "./utils";
  * An abstract component that Blueprint components can extend
  * in order to add some common functionality like runtime props validation.
  */
-export abstract class AbstractComponent<P, S> extends React.Component<P, S> {
+export abstract class AbstractComponent<P, S, SS = {}> extends React.Component<P, S, SS> {
     /** Component displayName should be `public static`. This property exists to prevent incorrect usage. */
     protected displayName: never;
 
     // Not bothering to remove entries when their timeouts finish because clearing invalid ID is a no-op
     private timeoutIds: number[] = [];
 
-    constructor(props?: P, context?: any) {
+    protected constructor(props?: P, context?: any) {
         super(props, context);
         if (!isNodeEnv("production")) {
             this.validateProps(this.props);
         }
     }
 
-    public componentWillReceiveProps(nextProps: P & { children?: React.ReactNode }) {
```

Comment:
I think we'll need to keep this method around for backwards compatibility, since `AbstractComponent` is exported from `@blueprintjs/core`. Its implementation can stay as-is, just marked as `@deprecated` in a JSDoc comment.

Reply:
after looking at the react-lifecycles-compat documentation, it seems like it will throw an error if `componentWillReceiveProps` is defined. So I propose that we keep around the old `AbstractComponent` but make a new version of this class with a new name which includes the changes in this diff and is used internally in Blueprint components with `polyfill()`

Reply:
i've added `UNSAFE_` components for the abstract components as well as the ones that implement them. is that what you had in mind?